examples/probability.dx

```diff
@@ -0,0 +1,517 @@
+  You are asked if you want to change your selection...
+
+Doors = Fin 3
+YesNo = { no:Unit | yes:Unit}
+def yesno (x:Bool) : Dist YesNo = delta $ select x {|yes=()|} {|no=()|}
+
+' The generative model is relatively simple
+  1. We will first sample our pick and the door.
+  1. Then we will consider changing our pick.
+  1. Finally we will see if we won.
+
+
+def monty_hall (change': Var YesNo) (win': Var YesNo) : Float =
+    (one ~ uniform) (for (pick, correct): (Doors & Doors).
+     (change' ~ uniform) (for change.
+        win_dist = case change of 
+                     {|yes=()|} -> yesno (pick /= correct)
+                     {|no=()|}  -> yesno (pick == correct)
+        (win' ~ win_dist) one))
+
+' To check the odds we will compute probabity of winning conditioned
+  on changing.
+
+{|yes=()|} ?? (posterior $ monty_hall (observed {|yes=()|}))
+
+
+' And compare to proability of winning with no change.
+
+{|yes=()|} ?? (posterior $ monty_hall (observed {|no=()|}))
+
+
+
+' ## Example 5: Hidden Markov Models
+
+' Finally we conclude with a more complex example. A hidden Markov model is
+  one of the most widely used discrete time series models. It models the relationship between discrete hidden states $Z$ and emissions $X$.
+
+Z = Fin 5
+X = Fin 10
+
+' It consists of three distributions: initial, transition, and emission.
+
+initial : Pr Z nil = arb $ newKey 1
+emission : Pr X Z = arb $ newKey 2
+transition : Pr Z Z = arb $ newKey 3
+
+' The model itself takes the following form for $m$ steps.
+'
+  $$ z_0 \sim \text{initial}$$
+  $$ z_1 \sim \text{transition}(z_0)$$
+  $$ x_1 \sim \text{emission}(z_1)$$
+  $$ ...$$
+
+' This is implemented in reverse order for clarity (backward algorithm).
+
+def hmm (init': Var Z) (x': m => Var X) (z' : m => Var Z) : Float =
+    (init' ~ initial.nil) $ yieldState one ( \future .
+        for i:m.
+            j = ((size m) - (ordinal i) - 1)@_
+            future := for z.
+                (x'.j ~ emission.z) (for _.
+                 (z'.j ~ transition.z) (get future)))
+
+
+' We can marginalize out over latents.
+
+hmm (observed (1@_)) (for i:(Fin 2). observed (1@_)) (for i. latent)
+
+
+' Or we can compute the posterior probabilities of specific values.
+
+posteriorTab $ \z . hmm (observed (1@_)) (for i:(Fin 2). observed (1@_)) z
+
+' ## Example 5a. HMM Monoid 
+
+' We can also write out an HMM using a Monoid. Here we define a monoid
+  for square matrix multiplication.
+
+def MarkovMonoid (a:Type) : Monoid (a => a => Float) =
+  M = a -- XXX: Typing `Monoid a` below would quantify it over a, which we don't want
+  named-instance result : Monoid (M => M => Float)
+     mempty = for m1 m2. select ((ordinal m1) == (ordinal m2)) 1.0 0.0
+     mcombine = \m1 m2. for i j. sum for k. m1.i.k * m2.k.j
+  result
+
+' We also define a Markov version of our sample function.
+  Instead of summing out over the usage of its result,
+  it constructs a matrix a vector.
```

Comment:
I can't follow this

Reply:
Will make more clear. This part was me trying to follow @duvenaud 's particle filter and rewrite it as an associative Accum.